components/application-mgt/org.wso2.carbon.identity.application.mgt/src/test/java/org/wso2/carbon/identity/application/mgt/ApplicationManagementServiceImplTest.java

```diff
@@ -1309,6 +1346,251 @@ public void addApplicationWithTemplateIdAndTemplateVersionData(String templateId
     }
 
+    @Test(groups = "certificate", priority = 1)
+    public void testAddApplicationWithCertificate() throws CertificateMgtException {
+
+        ServiceProvider inputSP = new ServiceProvider();
+        inputSP.setApplicationName(APPLICATION_NAME_1);
+        inputSP.setCertificateContent(certificate.getCertificateContent());
+        addApplicationConfigurations(inputSP);
+
+        reset(applicationCertificateManagementService);
+        doReturn(0).when(applicationCertificateManagementService).addCertificate(any(), anyString());
+        doReturn(certificate).when(applicationCertificateManagementService).getCertificateByName(any(), anyString());
+        try {
+            appIdWithCert = applicationManagementService.createApplication(inputSP, SUPER_TENANT_DOMAIN_NAME,
+                    REGISTRY_SYSTEM_USERNAME);
+        } catch (Exception e) {
+            Assert.fail();
+        }
+    }
+
+    @Test(groups = "certificate", priority = 2, dependsOnMethods = "testAddApplicationWithCertificate")
+    public void testGetApplicationWithServerErrorWhenRetrievingCert() throws CertificateMgtException {
+
+        reset(applicationCertificateManagementService);
+        doThrow(serverException).when(applicationCertificateManagementService).getCertificate(anyInt(), anyString());
+        try {
+            applicationManagementService.getApplicationByResourceId(appIdWithCert, SUPER_TENANT_DOMAIN_NAME);
+            Assert.fail();
+        } catch (IdentityApplicationManagementException e) {
+            Assert.assertEquals(e.getClass(), IdentityApplicationManagementServerException.class);
+            Assert.assertTrue(e.getMessage().contains("Error while retrieving application with resourceId: "
+                    + appIdWithCert + " in tenantDomain: " + SUPER_TENANT_DOMAIN_NAME));
+            for (Throwable cause = e.getCause(); cause != null; cause = cause.getCause()) {
+                if (cause instanceof CertificateMgtServerException) {
+                    return;
+                }
+            }
+            Assert.fail();
+        }
+    }
+
+    @Test(groups = "certificate", priority = 3, dependsOnMethods = "testAddApplicationWithCertificate")
+    public void testGetApplicationWithCertificate() throws IdentityApplicationManagementException,
+            CertificateMgtException {
+
+        reset(applicationCertificateManagementService);
+        when(applicationCertificateManagementService.getCertificate(anyInt(), anyString())).thenReturn(certificate);
+        appWithCert = applicationManagementService.getApplicationByResourceId(appIdWithCert, SUPER_TENANT_DOMAIN_NAME);
+
+        String certificateValue = Arrays.stream(appWithCert.getSpProperties())
+                .filter(prop -> "CERTIFICATE".equals(prop.getName()))
+                .map(ServiceProviderProperty::getValue)
+                .findFirst()
+                .orElseThrow(() -> new AssertionError("Certificate property not found"));
```

Comment:
can't we just have orElse and invoke assert.fail ?

Reply:
We can't invoke `Assert.fail(`) inside `.orElse()` since, `Assert.fail()` does not return a value. But `.orElse()` expects a default value of the same type as the other path, which in this case would be a `String`.

Reply:
But improved it without using AssertionError.
https://github.com/wso2/carbon-identity-framework/pull/6087/commits/d3a10433e71b9f0976aef3b255ae4432ffb04470